### PR TITLE
Fix/55278 standardize font library rest api

### LINKED ISSUE #55278: Font Library: refactor REST API
## What?
Font Library: refactor REST API endpoints.

### Refactor REST endpoint URLs
([reference](https://github.com/WordPress/wordpress-develop/pull/5285#discussion_r1335911520) [reference](https://github.com/WordPress/wordpress-develop/pull/5285#discussion_r1335916249)).
PR:
- https://github.com/WordPress/gutenberg/pull/57282

### Transition the Font Family APIs to be SINGULAR
(accepting as parameters and returning as responses a single Font Family item) ([reference](https://core.trac.wordpress.org/ticket/59166#comment:9) [reference](https://wordpress.slack.com/archives/C02RQC26G/p1696281308965689) )

PR:
 - https://github.com/WordPress/gutenberg/pull/57569

### Return a single Error (with multiple calls to `add()`)
instead of a collection of errors. ([reference](https://github.com/WordPress/wordpress-develop/pull/5285#discussion_r1337045813) [reference](https://github.com/WordPress/wordpress-develop/pull/5285#discussion_r1337060481) )

### Return API responses with `rest_ensure_response` instead of `new WP_REST_Response()`
([reference](https://github.com/WordPress/wordpress-develop/pull/5285#discussion_r1342760618))

### Ensure there is error handling of any decoded JSON string
( [reference](https://github.com/WordPress/wordpress-develop/pull/5285#discussion_r1342767882) )

### Add schema's to the requests
([reference](https://github.com/WordPress/wordpress-develop/pull/5285#discussion_r1342770186))

### Break the font library REST API controller in 2 controllers.
One for font families and one for font collections. ([reference](https://wordpress.slack.com/archives/C02RQC26G/p1696366631579099))

PR:
- https://github.com/WordPress/gutenberg/pull/57282

### Error code strings need to be translated
([[reference](https://github.com/WordPress/wordpress-develop/pull/5285/files#r1345355956)]). 

### Ensure the gutenberg domain used for translations is removed when porting to core.
([reference](https://github.com/WordPress/wordpress-develop/pull/5285#discussion_r1335914971))

## Why?
To standardize the endpoint routes and input/output data formats and schemas.

## How?
This task is somewhat big, so multiple PRs for it are expected.

--- 

Props to all the folks that brought this issue in this PR: https://github.com/WordPress/wordpress-develop/pull/5285

### PR DESCRIPTION
**Pull request by pbking**
## What?
Fixes #55278

* Refactors the Font Library REST URLs to use more standard locations
* Refactors the Font Library REST endpoints to use SINGULAR items instead of PLURAL items
* Uses `rest_ensure_response` for all API responses
* Included error handling of any JSON decoding
* adds schemas to the APIs
* 1 controller for Font Families, 1 for Font Collections
* Translate all error codes
* View changes to support SINGULAR responses and new error responses.

## Why?
The original implementation was "non standard" in a number of ways (documented in #55278)

## How?
Much of the work was previously accomplished during the [attempt to merge](https://github.com/WordPress/wordpress-develop/pull/5285) this work into WordPress Core for the 6.4 release.  Those changes have been ported here (as well as the view changes necessary).

## Testing Instructions
* Upload a Font via the upload UI.  It should be activated and useable on your site.
* Upload multiple fonts at a time.  They should all be available.
* Attempt to upload fonts from different families at the same time.  It should fail with an error message.
image
Delete the uploaded font(s).  It should be removed from the system and no longer available.
image
Install a font from Google Fonts. It should be activated and useable on your site.
image

### CODE REVIEW

File: lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+	}
+
+	/**
+	 * Checks whether the user has permissions to update the Font Library.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return true|WP_Error True if the request has write access for the item, WP_Error object otherwise.
+	 */
+	public function update_font_library_permissions_check() {
```
Comment (danielbachhuber) [inline lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php]: If we only have `WP_REST_Server::READABLE` endpoints here, this should be `get_items_permissions_check()`, e.g.

https://github.com/WordPress/wordpress-develop/blob/6528d9840b6800edfb8e75c4924936bb14b3f31b/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L146-L168

File: lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+		if ( ! current_user_can( 'edit_theme_options' ) ) {
+			return new WP_Error(
+				'rest_cannot_update_font_library',
+				__( 'Sorry, you are not allowed to update the Font Library on this site.' ),
```
Comment (danielbachhuber) [inline lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php]:
```suggestion
				__( 'Sorry, you are not allowed to get the Font Library on this site.' ),
```

File: lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+	 * Checks whether the user has permissions to update the Font Library.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return true|WP_Error True if the request has write access for the item, WP_Error object otherwise.
+	 */
+	public function update_font_library_permissions_check() {
+		if ( ! current_user_can( 'edit_theme_options' ) ) {
+			return new WP_Error(
+				'rest_cannot_update_font_library',
```
Comment (danielbachhuber) [inline lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php]:
```suggestion
				'rest_cannot_get_font_library',
```

File: phpunit/tests/fonts/font-library/wpRestFontCollectionController/wpRestFontCollectionsController.php [hunk 4]
```diff
@@ -0,0 +1,300 @@
+<?php
```
Comment (danielbachhuber) [inline phpunit/tests/fonts/font-library/wpRestFontCollectionController/wpRestFontCollectionsController.php]: This file needs some tests for the unauthenticated and unauthorized scenarios.

File: phpunit/tests/fonts/font-library/wpRestFontCollectionController/wpRestFontCollectionsController.php [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+/**
+ * Unit tests covering WP_REST_Font_Collections_Controller functionality.
+ *
+ * @package WordPress
+ * @subpackage REST API
+ *
+ * @group restapi
+ */
+class Tests_Fonts_WpRestFontCollectionsController extends WP_Test_REST_Controller_Testcase {
```
Comment (danielbachhuber) [inline phpunit/tests/fonts/font-library/wpRestFontCollectionController/wpRestFontCollectionsController.php]: I'd expect a class name similar to `WP_Test_REST_Posts_Controller`. Is there a reason you're using camelcase here?

File: phpunit/tests/fonts/font-library/wpRestFontFamiliesController/deleteItem.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * @package WordPress
+ * @subpackage Fonts
+ *
+ * @group fonts
+ * @group font-library
+ *
+ * @covers WP_REST_Font_Families_Controller::delete_item
+ */
+class Tests_Fonts_WPRESTFontFamiliesController_DeleteItem extends WP_REST_Font_Families_Controller_UnitTestCase {
```
Comment (danielbachhuber) [inline phpunit/tests/fonts/font-library/wpRestFontFamiliesController/deleteItem.php]: Can we consolidate these `Tests_Fonts_WPRESTFontFamiliesController_*` test classes into one larger `WP_Test_REST_Font_Families_Controller` class?

File: lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+	 * to the Font Library.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param WP_REST_Request $request The request object containing the new fonts to install
+	 *                                 in the request parameters.
+	 * @return WP_REST_Response|WP_Error The updated Font Library post content.
+	 */
+	public function create_item( $request ) {
+		$font_family_data = array(
```
Comment (danielbachhuber) [inline lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php:321]: What sort of validation/sanitization do these request arguments need?

File: lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php [hunk 8]
```diff
@@ -0,0 +1,503 @@
...
+			'slug'       => $request->get_param( 'slug' ),
+			'name'       => $request->get_param( 'name' ),
+			'fontFamily' => $request->get_param( 'fontFamily' ),
```
Comment (danielbachhuber) [inline lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php]: All request arguments should be snake case, not camelcase

File: lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php [hunk 9]
```diff
@@ -0,0 +1,503 @@
...
+		);
+
+		if ( $request->get_param( 'fontFace' ) ) {
+			$font_family_data['fontFace'] = json_decode( $request->get_param( 'fontFace' ), true );
```
Comment (danielbachhuber) [inline lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php]: What sort of validation/sanitization does this need?

**Review by danielbachhuber — COMMENTED** [review]
Some initial thoughts based on a cursory review

File: lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php (hunk 3 above)
Comment (pbking) [reply lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php]: changed

File: lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php (hunk 2 above)
Comment (pbking) [reply lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php]: 👍

File: lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php (hunk 1 above)
Comment (pbking) [reply lib/experimental/fonts/font-library/class-wp-rest-font-collections-controller.php]: 👍

File: lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php (hunk 8 above)
Comment (pbking) [reply lib/experimental/fonts/font-library/class-wp-rest-font-families-controller.php]: Updated.  
Note that data being passed around is still camelCase (so as to align with theme.json style data structure) however the API arguments are now all snake case.

File: phpunit/tests/fonts/font-library/wpRestFontFamiliesController/deleteItem.php (hunk 6 above)
Comment (pbking) [reply phpunit/tests/fonts/font-library/wpRestFontFamiliesController/deleteItem.php]: I suppose we could... what's the benefit? The `create_item` tests especially are pretty big and combining them into a single file seems unnecessary.  Other tests in the suite are broken similarly.  Would you want to consider consolidation of those too?

File: phpunit/tests/fonts/font-library/wpRestFontCollectionController/wpRestFontCollectionsController.php (hunk 5 above)
Comment (pbking) [reply phpunit/tests/fonts/font-library/wpRestFontCollectionController/wpRestFontCollectionsController.php]: Renamed.  I think it's closer to what's expected now.

**Comment by pbking** [comment]
closing in leu of #57688

**Pull request closed without merging**